diff --git a/source/linux-4.14.290/drivers/net/phy/realtek.c b/source/linux-4.14.290/drivers/net/phy/realtek.c
old mode 100644
new mode 100755
index 7d38af5ed..54e52d352
--- a/source/linux-4.14.290/drivers/net/phy/realtek.c
+++ b/source/linux-4.14.290/drivers/net/phy/realtek.c
@@ -107,7 +107,8 @@ static int rtl8211f_config_init(struct phy_device *phydev)

        /* enable TX-delay for rgmii-id and rgmii-txid, otherwise disable it */
        if (phydev->interface == PHY_INTERFACE_MODE_RGMII_ID ||
-           phydev->interface == PHY_INTERFACE_MODE_RGMII_TXID)
+           phydev->interface == PHY_INTERFACE_MODE_RGMII_TXID ||
+           phydev->interface == PHY_INTERFACE_MODE_RGMII)
                reg |= RTL8211F_TX_DELAY;
        else
                reg &= ~RTL8211F_TX_DELAY;
@@ -168,7 +169,7 @@ static struct phy_driver realtek_drvs[] = {
                .phy_id         = 0x001cc916,
                .name           = "RTL8211F Gigabit Ethernet",
                .phy_id_mask    = 0x001fffff,
-               .features       = PHY_GBIT_FEATURES,
+               .features       = PHY_BASIC_FEATURES,
                .flags          = PHY_HAS_INTERRUPT,
                .config_aneg    = &genphy_config_aneg,
                .config_init    = &rtl8211f_config_init,

